packages/common/src/type-helpers.ts

```diff
@@ -141,8 +141,35 @@ export function SymbolBasedInstanceOfError<E extends Error>(markerName: string):
   };
 }
 
-// Thanks MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze
-export function deepFreeze<T>(object: T): T {
+// This implementation is based on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze.
+// Note that there are limits to this approach, as traversing using getOwnPropertyXxx doesn't allow
+// reaching variables defined using internal scopes. This implementation specifically look for and deal
+// with the case of the Map or Set classes, but it is impossible to cover all potential cases.
+export function deepFreeze<T>(object: T, visited = new WeakSet<any>(), path = 'root'): T {
+  if (typeof object !== 'object' || object == null || Object.isFrozen(object) || visited.has(object)) return object;
+  visited.add(object);
+
+  if (object.constructor?.name === 'Map' || object instanceof Map) {
+    const map = object as unknown as Map<any, any>;
+    for (const [k, v] of map.entries()) {
+      deepFreeze(k, visited, `${path}[key:${k}]`);
+      deepFreeze(v, visited, `${path}[value:${k}]`);
+    }
+    map.set = frozenMapSet;
+    map.delete = frozenMapDelete;
+    map.clear = frozenMapClear;
```

Comment:
Do you need to freeze the map itself too so that they don't override .set .delete .clear again?

Reply:
It is. After replacing these methods, we continue with the normal deepfreeze logic below, which will freeze the set/delete/clear methods, then the map itself.